janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStrategy.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.tinkerpop.gremlin.structure.T;
+import org.janusgraph.core.JanusGraphTransaction;
+import org.janusgraph.core.MixedIndexCountQuery;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.tinkerpop.optimize.HasStepFolder.validFoldInHasContainer;
+
+/**
+ * If the query can be satisfied by a single mixed index query, and the query is followed by a count step, then
+ * this strategy replaces original step with {@link JanusGraphMixedIndexCountStep}, which fires a count query against
+ * mixed index backend without retrieving all elements
+ *
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
+public class JanusGraphMixedIndexCountStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy>
+    implements TraversalStrategy.ProviderOptimizationStrategy {
+    private static final JanusGraphMixedIndexCountStrategy INSTANCE = new JanusGraphMixedIndexCountStrategy();
+
+    private JanusGraphMixedIndexCountStrategy() {
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (TraversalHelper.onGraphComputer(traversal))
+            return;
+
+        TraversalHelper.getStepsOfClass(GraphStep.class, traversal).forEach(originalGraphStep -> {
+            if (originalGraphStep.getIds() == null || originalGraphStep.getIds().length == 0) {
+                buildMixedIndexCountStep(originalGraphStep, traversal);
+            }
+        });
+    }
+
+    private void buildMixedIndexCountStep(final GraphStep originalGraphStep, final Traversal.Admin<?, ?> traversal) {
+        if (!hasCountGlobalStep(originalGraphStep)) {
+            return;
+        }
+        final JanusGraphMixedIndexCountStep<?> directQueryCountStep = new JanusGraphMixedIndexCountStep<>(traversal);
+        // we shouldn't really fold in, but only check it, because we don't want to break the original traversal,
+        // as this optimizer might fail
+        if (!foldInHasContainer(originalGraphStep, directQueryCountStep, traversal, traversal)) {
+            return;
+        }
+        final JanusGraphTransaction tx = JanusGraphTraversalUtil.getTx(traversal);
+        MixedIndexCountQuery mixedIndexCountQuery = directQueryCountStep.buildGlobalGraphIndexQuery(originalGraphStep, tx);
+        if (mixedIndexCountQuery != null) {
+            directQueryCountStep.setMixedIndexCountQuery(mixedIndexCountQuery);
+            removeHasContainerSteps(originalGraphStep, directQueryCountStep, traversal);
+        }
+    }
+
+    private boolean hasCountGlobalStep(final GraphStep originalGraphStep) {
+        Step<?, ?> currentStep = originalGraphStep.getNextStep();
+        while (currentStep instanceof IdentityStep || currentStep instanceof NoOpBarrierStep || currentStep instanceof HasContainerHolder) {
```

Comment:
These `instanceof` chains are used quite often so it would be better to move the check to a separate method. This would also give these steps a name that is more readable. For example `while (isEligibleToSkip(currentStep)) {...}`